### PR TITLE
[TT-311] Don't cache tls config  when we use certificate ID and redis is down.

### LINKED ISSUE #3292: Gateway fails to read certificate
**Branch/Environment/Version**
- Branch/Version: Stable 3.0
- Environment: All

**Describe the bug**
Gateway fails to read a certificate that was functional in tyk gateway 2.9.4

**Reproduction steps**

1. Add a .pem containing cert and private key to dashboard
2. Add the cert id to tyk.conf at http_server_options.ssl_certificates and set use_ssl true

**Actual behavior**
Gateway fails with the error:
`TLS handshake error from xxx.xxx.xxx.xxx:xxxxx: tls: no certificates configured`

Gateway additionally reports:
`level=warning msg="Reconnecting storage: Redis is either down or ws not configured" prefix=pub-sub`
But the certificate is visible in redis.

This behavior is observed both on a fresh installation of 3.0 connecting to a fresh docdb and redis cluster as well as upgrading and keeping the pre-existing data.

**Expected behavior**
Gateway starts correctly and receives ssl traffic from dashboard

### PR DESCRIPTION
**Pull request by gernest**
## Description

This  makes sure we don't cache tls configuration when we are expecting certificates from redis (via certificate ID) and redis is not available yet. The resulting tls handshake will fail, it will eventually pass when redis is available.

When certificates are added to dashboard, we make sure they are eventually loaded to the gateway.

## Related Issue
Fixes #3292

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

File: gateway/cert.go [hunk 1]
```diff
@@ -271,38 +272,40 @@ var tlsConfigCache = cache.New(60*time.Second, 60*time.Minute)
...
-		certNameMap[certData.Name] = &cert
-	}
 
-	for _, cert := range CertificateManager.List(config.Global().HttpServerOptions.SSLCertificates, certs.CertificatePrivate) {
-		if cert != nil {
-			serverCerts = append(serverCerts, *cert)
-		}
-	}
+		// if we have certificates ID's we don't cache the config if redis is not available yet.
+		noCaching := len(config.Global().HttpServerOptions.SSLCertificates) > 0 && !storage.Connected()
```
Comment (buger) [inline gateway/cert.go:283]: Should not depend just on `SSLCertificates` field. If you use mTLS or similar you also should not cache if redis not enabled. So just change to `storage.Connected()`

Comment (buger) [reply gateway/cert.go:283]: Also add a comment with explanation why it is needed

**Comment by buger** [comment]
Needs JIRA ticket and test added

**Comment by sedkis** [comment]
For TT-311

**Comment by sredxny** [comment]
@gernest hi, is this PR ready for review? I see it as a draft

**Comment by gernest** [comment]
@sredxny this is ancient can't remember much. You can resurrect this if you want, or come up with better solution. I should probably close this.

**Pull request closed without merging**